### ISSUE TITLE
[Feature Request] Extend [modify_unit_type] tag with effect WML API

### ISSUE THREAD

**Issue by czyzby** [wesnoth/wesnoth#6446]
`[modify_unit_type]` tag currently allows changing some basic unit data, mostly related to the advancements. While useful for extending advancement trees of the existing units, it does not cover many use cases related to add-ons such as graphic packs or balancing patches. For example, it seems that it is currently not possible to effectively replace the sprites, animations, or portraits of a specific unit type both during a match, as well as in the in-game wiki.

The goal of this feature request is a partial or full support of the unit modifications provided by the [`[effect]`](https://wiki.wesnoth.org/EffectWML) tag including:
- Modification of unit statistics.
- Replacement of unit graphics and descriptions.
- Modification, addition or removal of unit attacks and abilities.

**Comment by CelticMinstrel** [comment]
How would this method be better than creating an entirely new unit type using `[base_unit]`?

**Comment by czyzby** [comment]
@CelticMinstrel Creating a new unit is fine if you're aiming for an era or a campaign, but not so much if you're trying to write a modification that is supposed to work with the mainline units.

All of these things that I've mentioned are possible to achieve by extending or copying the existing units, but implementing these like that just leads to add-ons that are incompatible with each other, and a lot of duplicated content. Ideally, things like slightly modifying existing units, factions or eras should be possible without having to add entirely new ones. See #6424 for a related issue.

Let's say I wanted to create a [modification that extends original Saurian units](https://github.com/czyzby/wesnoth-saurian-pack) with the following:
- Additional upgrades of the existing units.
- Alternative portraits for selected units.
- New leaders for the official Drakes factions.
- Upgraded leaders and new recruits in the Age of Heroes eras.

The upgrades can be easily added with `[modify_unit_type]`, which also happens to work as a `[resource]` that can be included in eras or campaigns. As far as I can tell, modifying portraits in a way that would work with the in-game wiki is not possible, or at least not with pure WML. Same goes for the new leaders. I think the recruits list can be extended through a combination of Lua scripts and WML, but the solution is quite hacky and requires quite a bit of setup for such a common use case.

Of course, I could create a new `Default + Saurians` era, or something along these lines. However, it's not nearly as reusable as a modification that would work with pretty much _every_ era and campaign that uses the default Saurians. In fact, it would lock the users into a specific era that would probably see barely any play. I'm guessing the modification would be pulled for the sprites more often than for the actual matches.

With what I'm proposing, it would be possible to write tiny modifications that are all compatible with the default era _and_ with each other. Say, something like balancing patches or alternative unit graphics.

**Comment by CelticMinstrel** [comment]
Only two out of your four bullet points could be done with `[modify_unit_type]` even with this change. And additional upgrades are already possible, since that's the use-case `[modify_unit_type]` was originally made for.

**Comment by czyzby** [comment]
Yeah, and the rest are covered by the issues that I have submitted. Even the relatively few features that my tiny add-on requires are not fully covered by WML right now, as far as I can tell. For example, I have not found a way to replace portrait of a unit type that would work with the in-game wiki.

Besides, is there any particular reason why `[modify_unit_type]` _shouldn't_ support changing statistics at the level of the unit type? Currently available tools often allow modifying these at the individual unit level, which is fine for scenarios and campaigns (e.g., to buff up specific units), but modifying "global" settings of specific unit types is a common use case for eras and modifications. And I'm not entirely sure if the current approach of copying or extending existing content is the best one, as it can lead to incompatibilities between add-ons that could very well be used together when reduced to the actual new features. As an example, all these `Default + X` eras could be mods that simply add another faction to the Default era, possibly with some small balancing updates.

**Comment by gfgtdf** [comment]
>

**Comment by gfgtdf** [comment]
Wrong button clicked
